File: 04-estructuras-datos/12-indices-y-organizacion-registros/java/ejemplos/IndicePorClave.java
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class IndicePorClave {
    record Cliente(String dni, String nombre) {}

    public static void main(String[] args) {
        List<Cliente> archivo = List.of(new Cliente("100", "Ana"), new Cliente("200", "Luis"));
        Map<String, Integer> indice = new HashMap<>();

        // El índice guarda la posición del registro asociada a su clave.
        for (int i = 0; i < archivo.size(); i++) indice.put(archivo.get(i).dni(), i);

        int posicion = indice.get("200");
        System.out.println(archivo.get(posicion));
    }
}

File: 04-estructuras-datos/12-indices-y-organizacion-registros/scala/ejemplos/IndicePorClave.scala
case class Cliente(dni: String, nombre: String)

object IndicePorClave:
  def main(args: Array[String]): Unit =
    val archivo = Vector(Cliente("100", "Ana"), Cliente("200", "Luis"))
    // zipWithIndex permite construir un índice clave -> posición.
    val indice = archivo.zipWithIndex.map((cliente, pos) => cliente.dni -> pos).toMap
    println(archivo(indice("200")))
